File: BubblerApp/BubblerApp/AuthSession.swift
//
//  AuthSession.swift
//  BubblerApp
//

import Combine
import Foundation

@MainActor
final class AuthSession: ObservableObject {
    @Published private(set) var user: BackendUser?
    @Published var authError: String?
    @Published var successMessage: String?
    @Published var isWorking = false
    @Published private(set) var isRestoringSession = false

    private let client: BackendClient
    private let storage: UserDefaults
    private let sessionUserKey = "bubbler.sessionUser"

    var isSignedIn: Bool {
        user != nil
    }

    init(client: BackendClient = .shared, storage: UserDefaults = .standard) {
        self.client = client
        self.storage = storage
        self.user = Self.storedUser(from: storage, key: sessionUserKey)
    }

    func signIn(email: String, password: String) async {
        let trimmedEmail = normalizedEmail(email)

        guard !trimmedEmail.isEmpty else {
            authError = "Enter your email address."
            return
        }

        guard !password.isEmpty else {
            authError = "Enter your password."
            return
        }

        _ = await performAuthAction {
            saveSession(try await client.loginSession(email: trimmedEmail, password: password))
        }
    }

    func createAccount(email: String, password: String, confirmPassword: String) async {
        let trimmedEmail = normalizedEmail(email)

        guard !trimmedEmail.isEmpty else {
            authError = "Enter your email address."
            return
        }

        guard password.count >= 6 else {
            authError = "Password must be at least 6 characters."
            return
        }

        guard password == confirmPassword else {
            authError = "Passwords do not match."
            return
        }

        let didCreateAccount = await performAuthAction {
            saveSession(try await client.registerSession(
                username: username(from: trimmedEmail),
                email: trimmedEmail,
                password: password
            ))
        }

        if didCreateAccount {
            successMessage = "Account created successfully!"
        }
    }

    func restoreSession() async {
        guard let token = SessionTokenStore.load() else {
            return
        }

        authError = nil
        isRestoringSession = true
        defer { isRestoringSession = false }

        do {
            saveSession(try await client.session(token: token))
        } catch {
            clearStoredSession()
        }
    }

    func signOut() {
        clearStoredSession()
        authError = nil
        successMessage = nil
    }

    func clearSuccessMessage() {
        successMessage = nil
    }

    private func performAuthAction(_ action: () async throws -> Void) async -> Bool {
        authError = nil
        successMessage = nil
        isWorking = true
        defer { isWorking = false }

        do {
            try await action()
            return true
        } catch {
            authError = error.localizedDescription
            return false
        }
    }

    private func normalizedEmail(_ email: String) -> String {
        email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    private func saveSession(_ response: AuthResponse) {
        user = response.user
        SessionTokenStore.save(response.sessionToken)

        if let encodedUser = try? JSONEncoder().encode(response.user) {
            storage.set(encodedUser, forKey: sessionUserKey)
        }
    }

    private func clearStoredSession() {
        user = nil
        SessionTokenStore.delete()
        storage.removeObject(forKey: sessionUserKey)
    }

    private static func storedUser(from storage: UserDefaults, key: String) -> BackendUser? {
        guard let data = storage.data(forKey: key) else {
            return nil
        }

        return try? JSONDecoder().decode(BackendUser.self, from: data)
    }

    private func username(from email: String) -> String {
        let prefix = email.split(separator: "@").first.map(String.init) ?? "bubbler"
        let safeCharacters = prefix.map { character in
            character.isLetter || character.isNumber ? character : "_"
        }

        let username = String(safeCharacters).trimmingCharacters(in: CharacterSet(charactersIn: "_"))
        return username.isEmpty ? "bubbler" : username
    }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/core/auth/AuthSession.kt
package com.bubbler.android.core.auth

import com.bubbler.android.core.network.ApiException
import com.bubbler.android.data.repository.AuthRepository
import com.bubbler.android.data.repository.AuthResponse
import com.bubbler.android.data.repository.PreferencesRepository
import com.bubbler.android.data.repository.UserRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.LocalDate
import java.time.Period
import java.util.Base64

object AgeGate {
    /** Global floor (EU-friendly GDPR Art. 8). */
    const val MINIMUM_AGE: Int = 16

    fun age(from: LocalDate, on: LocalDate = LocalDate.now()): Int =
        Period.between(from, on).years

    fun isOldEnough(dateOfBirth: LocalDate, on: LocalDate = LocalDate.now()): Boolean =
        age(from = dateOfBirth, on = on) >= MINIMUM_AGE

    val underageMessage: String
        get() = "You must be at least $MINIMUM_AGE years old to use Bubbler."
}

enum class OnboardingGate {
    Unknown,
    Required,
    Complete,
}

/**
 * Session gate matching iOS AuthSession: restore token, sign-in/register, sign-out.
 */
class AuthSession(
    private val tokenStore: TokenStore,
    private val authRepository: AuthRepository,
    private val userRepository: UserRepository? = null,
    private val preferencesRepository: PreferencesRepository? = null,
) {
    private val _accessToken = MutableStateFlow(tokenStore.loadAccessToken())
    val accessToken: StateFlow<String?> = _accessToken.asStateFlow()

    private val _userId = MutableStateFlow(restoredUserId(from = _accessToken.value))
    val userId: StateFlow<Int?> = _userId.asStateFlow()

    private val _isStaff = MutableStateFlow(false)
    val isStaff: StateFlow<Boolean> = _isStaff.asStateFlow()

    private val _authError = MutableStateFlow<String?>(null)
    val authError: StateFlow<String?> = _authError.asStateFlow()

    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()

    private val _isWorking = MutableStateFlow(false)
    val isWorking: StateFlow<Boolean> = _isWorking.asStateFlow()

    private val _onboardingGate = MutableStateFlow(OnboardingGate.Unknown)
    val onboardingGate: StateFlow<OnboardingGate> = _onboardingGate.asStateFlow()

    val isSignedIn: Boolean
        get() = _accessToken.value != null

    suspend fun signIn(email: String, password: String) {
        val trimmedEmail = normalizedEmail(email)

        when {
            trimmedEmail.isEmpty() -> {
                _authError.value = "Enter your email address."
                return
            }
            password.isEmpty() -> {
                _authError.value = "Enter your password."
                return
            }
        }

        performAuthAction(unauthorizedErrorMessage = "Incorrect username or password.") {
            authRepository.login(email = trimmedEmail, password = password)
        }
    }

    suspend fun createAccount(
        username: String,
        email: String,
        password: String,
        confirmPassword: String,
        dateOfBirth: LocalDate,
    ) {
        val trimmedUsername = username.trim()
        val trimmedEmail = normalizedEmail(email)

        when {
            trimmedUsername.isEmpty() -> {
                _authError.value = "Enter a username."
                return
            }
            trimmedUsername.length > 20 -> {
                _authError.value = "Username must be 20 characters or fewer."
                return
            }
            trimmedEmail.isEmpty() -> {
                _authError.value = "Enter your email address."
                return
            }
            !AgeGate.isOldEnough(dateOfBirth) -> {
                _authError.value = AgeGate.underageMessage
                return
            }
            password.length < 5 -> {
                _authError.value = "Password must be at least 5 characters."
                return
            }
            password.length > 40 -> {
                _authError.value = "Password must be 40 characters or fewer."
                return
            }
            password != confirmPassword -> {
                _authError.value = "Passwords do not match."
                return
            }
        }

        val didCreateAccount = performAuthAction {
            authRepository.register(
                username = trimmedUsername,
                email = trimmedEmail,
                password = password,
                dateOfBirth = dateOfBirth,
            )
        }

        if (didCreateAccount) {
            _successMessage.value = "Account created successfully!"
        }
    }

    suspend fun refreshStaffAccess() {
        if (!isSignedIn) {
            _isStaff.value = false
            return
        }
        val repo = userRepository ?: return
        try {
            _isStaff.value = repo.getProfile().isStaff
        } catch (e: ApiException.Unauthorized) {
            signOut()
        } catch (_: Exception) {
            // Keep the last known staff flag on transient failures.
        }
    }

    suspend fun refreshOnboardingStatus() {
        if (!isSignedIn) {
            _onboardingGate.value = OnboardingGate.Unknown
            return
        }
        val repo = preferencesRepository ?: run {
            _onboardingGate.value = OnboardingGate.Complete
            return
        }
        try {
            val preferences = repo.getPreferences()
            _onboardingGate.value = if (preferences.onboardingCompleted) {
                OnboardingGate.Complete
            } else {
                OnboardingGate.Required
            }
        } catch (e: ApiException.Unauthorized) {
            signOut()
        } catch (_: Exception) {
            _onboardingGate.value = OnboardingGate.Complete
        }
    }

    fun markOnboardingComplete() {
        _onboardingGate.value = OnboardingGate.Complete
    }

    fun signOut() {
        tokenStore.deleteAccessToken()
        _accessToken.value = null
        _userId.value = null
        _isStaff.value = false
        _onboardingGate.value = OnboardingGate.Unknown
        _authError.value = null
        _successMessage.value = null
    }

    fun clearSuccessMessage() {
        _successMessage.value = null
    }

    fun showSuccessMessage(message: String) {
        _successMessage.value = message
    }

    private suspend fun performAuthAction(
        unauthorizedErrorMessage: String? = null,
        action: suspend () -> AuthResponse,
    ): Boolean {
        _authError.value = null
        _successMessage.value = null
        _isWorking.value = true
        try {
            val response = action()
            tokenStore.saveAccessToken(response.accessToken)
            _accessToken.value = response.accessToken
            _userId.value = response.userId
            refreshStaffAccess()
            return true
        } catch (e: ApiException.Unauthorized) {
            _authError.value = unauthorizedErrorMessage ?: e.message
            return false
        } catch (e: Exception) {
            _authError.value = e.message ?: "Something went wrong."
            return false
        } finally {
            _isWorking.value = false
        }
    }

    private fun normalizedEmail(email: String): String =
        email.trim().lowercase()

    companion object {
        private val jwtJson = Json { ignoreUnknownKeys = true }

        fun restoredUserId(from: String?): Int? {
            if (from.isNullOrBlank()) return null

            val segments = from.split('.')
            if (segments.size < 2) return null

            var payload = segments[1]
                .replace('-', '+')
                .replace('_', '/')
            val padding = (4 - payload.length % 4) % 4
            if (padding > 0) {
                payload += "=".repeat(padding)
            }

            return try {
                val bytes = Base64.getDecoder().decode(payload)
                val root = jwtJson.parseToJsonElement(bytes.decodeToString()).jsonObject
                val subject = root["sub"]?.jsonPrimitive?.contentOrNull ?: return null
                subject.toIntOrNull()
            } catch (_: Exception) {
                null
            }
        }
    }
}
